<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from kinova.xacro                   | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="kinova">
  <!-- ========== -->
  <!--    LINKS   -->
  <!-- ========== -->
  <!-- ARM base link -->
  <link name="kinova_base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.00244324 0.00015573 0.08616742"/>
      <mass value="1.14608471"/>
      <inertia ixx="0.00335854" ixy="3.9E-07" ixz="0.00010989" iyy="0.003311" iyz="1.91E-06" izz="0.00077158"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/base_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/base_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- Shoulder link -->
  <link name="shoulder_link">
    <inertial>
      <origin rpy="0 0 0" xyz="2.477E-05 0.02213531 0.09937686"/>
      <mass value="0.95974404"/>
      <inertia ixx="0.00165947" ixy="2E-08" ixz="3.6E-07" iyy="0.00140355" iyz="0.00034927" izz="0.00089493"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/shoulder_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/shoulder_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- Arm link -->
  <link name="arm_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.02998299 0.21154808 0.0453031"/>
      <mass value="1.17756164"/>
      <inertia ixx="0.01149277" ixy="1E-06" ixz="1.6E-07" iyy="0.00102851" iyz="0.00140765" izz="0.01133492"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/arm_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/arm_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- Forearm link -->
  <link name="forearm_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.0301559 0.09502206 0.0073555"/>
      <mass value="0.59767669"/>
      <inertia ixx="0.00163256" ixy="7.11E-06" ixz="1.54E-06" iyy="0.00029798" iyz="9.587E-05" izz="0.00169091"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/forearm_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/forearm_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- Lower_wrist_link  -->
  <link name="lower_wrist_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.00575149 0.01000443 0.08719207"/>
      <mass value="0.52693412"/>
      <inertia ixx="0.00069098" ixy="2.4E-07" ixz="0.00016483" iyy="0.00078519" iyz="7.4E-07" izz="0.00034115"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/lower_wrist_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/lower_wrist_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- Upper_wrist_link -->
  <link name="upper_wrist_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.08056517 0.00980409 0.01872799"/>
      <mass value="0.58097325"/>
      <inertia ixx="0.00021268" ixy="5.21E-06" ixz="2.91E-06" iyy="0.00106371" iyz="1.1E-07" izz="0.00108465"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/upper_wrist_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/upper_wrist_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- Dummy links for transforming axes -->
  <link name="end_effector_link">
    <inertial>
      <mass value="0.0001"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0" iyy="0.0001" iyz="0.0" izz="0.0001"/>
    </inertial>
  </link>
  <link name="dummy_link"/>
  <link name="tool_frame"/>
  <!-- ================= -->
  <!-- Revolute JOINTS   -->
  <!-- ================= -->
  <!-- Joint 1 -->
  <joint name="joint_1" type="revolute">
    <origin rpy="0 0 0" xyz="0 0 0.12825"/>
    <parent link="kinova_base_link"/>
    <child link="shoulder_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10" lower="-2.68" upper="2.68" velocity="1.6"/>
  </joint>
  <!-- Joint 2 -->
  <joint name="joint_2" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0 -0.03 0.115"/>
    <parent link="shoulder_link"/>
    <child link="arm_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="14" lower="-2.61" upper="2.61" velocity="1.6"/>
  </joint>
  <!-- Joint 3 -->
  <joint name="joint_3" type="revolute">
    <origin rpy="-3.1416 0 0" xyz="0 0.28 0"/>
    <parent link="arm_link"/>
    <child link="forearm_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10" lower="-2.61" upper="2.61" velocity="1.6"/>
  </joint>
  <!-- Joint 4 -->
  <joint name="joint_4" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0 -0.14 0.02"/>
    <parent link="forearm_link"/>
    <child link="lower_wrist_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="39" lower="-2.57" upper="2.57" velocity="1.3963"/>
  </joint>
  <!-- Joint 5 -->
  <joint name="joint_5" type="revolute">
    <origin rpy="0 1.5708 0" xyz="0.0285 0 0.105"/>
    <parent link="lower_wrist_link"/>
    <child link="upper_wrist_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="7" lower="-2.53" upper="2.53" velocity="1.6"/>
  </joint>
  <!-- Joint 6 -->
  <joint name="joint_6" type="revolute">
    <origin rpy="0 -1.5708 0" xyz="-0.105 0 0.0285"/>
    <parent link="upper_wrist_link"/>
    <child link="end_effector_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="9" lower="-2.09" upper="2.09" velocity="1.2218"/>
  </joint>
  <!-- ================= -->
  <!-- Fixed JOINTS      -->
  <!-- ================= -->
  <joint name="end_effector" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="end_effector_link"/>
    <child link="dummy_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="0" upper="0" velocity="0"/>
  </joint>
  <!-- Tool frame used by the arm -->
  <joint name="tool_frame_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.130"/>
    <parent link="dummy_link"/>
    <child link="tool_frame"/>
    <axis xyz="0 0 0"/>
  </joint>
  <!-- Gripper description -->
  <!-- Gripper base link -->
  <link name="gripper_base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="4.03E-06 1.08E-05 0.041397"/>
      <mass value="0.1395"/>
      <inertia ixx="0.00011614" ixy="1E-08" ixz="4E-08" iyy="0.00010327" iyz="0" izz="9.707E-05"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/gripper_base_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/gripper_base_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- right_finger_prox_link -->
  <link name="right_finger_prox_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.020257 0.0063483 6.991E-05"/>
      <mass value="0.018385"/>
      <inertia ixx="2.29E-06" ixy="1.43E-06" ixz="0" iyy="4.56E-06" iyz="0" izz="4.77E-06"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/right_finger_prox_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/right_finger_prox_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- right_finger_dist_link -->
  <link name="right_finger_dist_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.018488 0.0011091 0"/>
      <mass value="0.010748"/>
      <inertia ixx="6.3E-07" ixy="3.2E-07" ixz="0" iyy="1.83E-06" iyz="0" izz="1.61E-06"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/right_finger_dist_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0 0.055 0.525 0.3"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/right_finger_dist_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- left_finger_prox_link -->
  <link name="left_finger_prox_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.020257 0.0063483 6.99E-05"/>
      <mass value="0.018385"/>
      <inertia ixx="2.29E-06" ixy="1.43E-06" ixz="0" iyy="4.56E-06" iyz="0" izz="4.77E-06"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/left_finger_prox_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.803 0.824 0.820 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/left_finger_prox_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- left_finger_dist_link -->
  <link name="left_finger_dist_link">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.018488 0.0011091 0"/>
      <mass value="0.010748"/>
      <inertia ixx="6.3E-07" ixy="3.2E-07" ixz="0" iyy="1.83E-06" iyz="0" izz="1.61E-06"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/left_finger_dist_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0 0.055 0.525 0.3"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/kinova/left_finger_dist_link.STL"/>
      </geometry>
    </collision>
  </link>
  <!-- ================= -->
  <!--        JOINTS     -->
  <!-- ================= -->
  <!-- right_finger_bottom_joint => THIS IS THE ONE THAT YOU ACTUALLY CONTROL, other joints mimic this one -->
  <joint name="right_finger_bottom_joint" type="fixed">
    <origin rpy="0 1.5708 0" xyz="0 -0.030501 0.070003"/>
    <parent link="gripper_base_link"/>
    <child link="right_finger_prox_link"/>
  </joint>
  <!-- right_finger_tip_joint -->
  <joint name="right_finger_tip_joint" type="fixed">
    <origin rpy="0 0 0" xyz="-0.045636 0.020423 0"/>
    <parent link="right_finger_prox_link"/>
    <child link="right_finger_dist_link"/>
  </joint>
  <!-- left_finger_bottom_joint  -->
  <joint name="left_finger_bottom_joint" type="fixed">
    <origin rpy="0 1.5708 0" xyz="0 0.0305 0.070003"/>
    <parent link="gripper_base_link"/>
    <child link="left_finger_prox_link"/>
  </joint>
  <!-- left_finger_tip_joint -->
  <joint name="left_finger_tip_joint" type="fixed">
    <origin rpy="0 0 0" xyz="-0.045636 -0.020423 6.9901E-05"/>
    <parent link="left_finger_prox_link"/>
    <child link="left_finger_dist_link"/>
  </joint>
  <joint name="arm_gripper_base_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="dummy_link"/>
    <child link="gripper_base_link"/>
  </joint>
  <!-- Create world joint -->
  <link name="world"/>
  <joint name="arm_mount_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="world"/>
    <child link="kinova_base_link"/>
  </joint>
  <link name="ball_link">
    <visual>
      <material name="red">
        <color rgba="1 0 0 1"/>
      </material>

      <geometry>
        <sphere radius="0.005"/>
      </geometry>
    </visual>
  </link>
  <joint name="ball_joint" type="fixed">
    <origin rpy="0.0 0.0 0.0" xyz="0.00038860649894460975 -0.0005822849938650045 0.180196205872143" />

    <parent link="gripper_base_link"/>
    <child link="ball_link"/>
  </joint>
</robot>

